crates/extrinsics/src/contract_storage.rs

```diff
@@ -0,0 +1,312 @@
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with cargo-contract.  If not, see <http://www.gnu.org/licenses/>.
+
+use anyhow::Result;
+use ink_metadata::{
+    layout::Layout,
+    InkProject,
+};
+use scale_info::form::PortableForm;
+use serde::Serialize;
+use sp_core::storage::ChildInfo;
+use std::{
+    collections::BTreeMap,
+    fmt::Display,
+};
+use subxt::{
+    backend::{
+        legacy::{
+            rpc_methods::Bytes,
+            LegacyRpcMethods,
+        },
+        rpc::{
+            rpc_params,
+            RpcClient,
+        },
+        BlockRef,
+    },
+    error::DecodeError,
+    ext::scale_decode::{
+        IntoVisitor,
+        Visitor,
+    },
+    Config,
+    OnlineClient,
+};
+
+use super::{
+    fetch_contract_info,
+    url_to_string,
+    Client,
+    ContractInfo,
+    DefaultConfig,
+    TrieId,
+};
+
+pub struct ContractStorage<C: Config = DefaultConfig> {
+    rpc: ContractStorageRpc<C>,
+}
+
+impl<C: Config> ContractStorage<C>
+where
+    C::AccountId: AsRef<[u8]> + Display + IntoVisitor,
+    DecodeError: From<<<C::AccountId as IntoVisitor>::Visitor as Visitor>::Error>,
+    BlockRef<sp_core::H256>: From<C::Hash>,
+{
+    pub fn new(rpc: ContractStorageRpc<C>) -> Self {
+        Self { rpc }
+    }
+
+    /// Load the raw key/value storage for a given contract.
+    pub async fn load_contract_storage_data(
+        &self,
+        contract_account: &C::AccountId,
+    ) -> Result<ContractStorageData> {
+        let contract_info = self.rpc.fetch_contract_info(contract_account).await?;
+        let trie_id = contract_info.trie_id();
+
+        let storage_keys = self
+            .rpc
+            .fetch_storage_keys_paged(trie_id, None, 1000, None, None) // todo loop pages
+            .await?;
+        let storage_values = self
+            .rpc
+            .fetch_storage_entries(trie_id, &storage_keys, None)
+            .await?;
+        assert_eq!(
+            storage_keys.len(),
+            storage_values.len(),
+            "storage keys and values must be the same length"
+        );
+        let storage = storage_keys
+            .into_iter()
+            .zip(storage_values.into_iter())
+            .filter_map(|(key, value)| value.map(|v| (key, v)))
+            .collect();
+
+        let contract_storage = ContractStorageData(storage);
+        Ok(contract_storage)
+    }
+
+    pub async fn load_contract_storage_with_layout(
+        &self,
+        metadata: &InkProject,
+        contract_account: &C::AccountId,
+    ) -> Result<ContractStorageLayout> {
+        let data = self.load_contract_storage_data(contract_account).await?;
+        let layout = ContractStorageLayout::new(data, metadata.layout());
+        Ok(layout)
+    }
+}
+
+/// Represents the raw key/value storage for the contract.
+#[derive(Serialize)]
+pub struct ContractStorageData(BTreeMap<Bytes, Bytes>);
```

Comment:
Why do you need a dedicated structure for this if it only contains a single field, and has no methods?